File: app/src/main/java/com/veyndan/paper/reddit/post/media/mutator/ImageMutatorFactory.java
package com.veyndan.paper.reddit.post.media.mutator;

import android.support.annotation.StringRes;
import android.util.Size;

import com.veyndan.paper.reddit.api.reddit.model.PostHint;
import com.veyndan.paper.reddit.api.reddit.model.Source;
import com.veyndan.paper.reddit.post.media.model.Image;
import com.veyndan.paper.reddit.post.model.Post;

import io.reactivex.Maybe;
import io.reactivex.Observable;
import io.reactivex.Single;

public final class ImageMutatorFactory implements MutatorFactory {

    public static ImageMutatorFactory create() {
        return new ImageMutatorFactory();
    }

    private ImageMutatorFactory() {
    }

    @Override
    public Maybe<Post> mutate(final Post post) {
        return Single.just(post)
                .filter(post1 -> post1.postHint() == PostHint.IMAGE)
                .map(post1 -> {
                    final boolean imageDimensAvailable = !post1.preview().images.isEmpty();

                    final Size size;
                    if (imageDimensAvailable) {
                        final Source source = post1.preview().images.get(0).source;
                        size = new Size(source.width, source.height);
                    } else {
                        size = new Size(0, 0);
                    }

                    @StringRes final int type = post1.linkUrl().endsWith(".gif")
                            ? Image.IMAGE_TYPE_GIF
                            : Image.IMAGE_TYPE_STANDARD;
                    final Image image = Image.create(post1.linkUrl(), size, type);
                    return post1.withMedias(post1.medias().value.concatWith(Observable.just(image)));
                });
    }
}

File: app/src/main/java/com/veyndan/paper/reddit/post/media/mutator/ImageMutatorFactory.kt
package com.veyndan.paper.reddit.post.media.mutator

import android.support.annotation.StringRes
import android.util.Size
import com.veyndan.paper.reddit.api.reddit.model.PostHint
import com.veyndan.paper.reddit.api.reddit.model.Source
import com.veyndan.paper.reddit.post.media.model.Image
import com.veyndan.paper.reddit.post.model.Post
import io.reactivex.Maybe
import io.reactivex.Observable
import io.reactivex.Single

class ImageMutatorFactory : MutatorFactory {

    override fun mutate(post: Post): Maybe<Post> {
        return Single.just(post)
                .filter { it.postHint == PostHint.IMAGE }
                .map {
                    val imageDimensAvailable: Boolean = it.preview.images.isNotEmpty()

                    val size = if (imageDimensAvailable) {
                        val source: Source = it.preview.images[0].source
                        Size(source.width, source.height)
                    } else {
                        Size(0, 0)
                    }

                    @StringRes val type: Int = if (it.linkUrl.endsWith(".gif")) Image.IMAGE_TYPE_GIF else Image.IMAGE_TYPE_STANDARD
                    val image: Image = Image(it.linkUrl, size, type)
                    it.copy(it.medias.concatWith(Observable.just(image)))
                }
    }
}
